waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,716 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Backend.ConcreteParser.Internal
+  ( -- * Internal parsing library
+    GrammarRule,
+    parse,
+
+    -- * Combinators
+
+    -- | The fixity of the operators ('(<>)', '`as`', '(<$$>)', and '(<|>)') was
+    -- chosen to make using operators require as little parentheses as possible:
+    --
+    -- >>> Program <$$> lparen <> rparen
+    --       <|> string
+    -- Alternative (Group Program (Chain (Token LParen) (Token RParen))) (Token String)
+    --
+    -- It's still recommended to use parentheses when white space does not make
+    -- the meaning extremely clear.
+    as,
+    (<$$>),
+    perhaps,
+    (<|>),
+
+    -- * Primitive grammar rules
+
+    -- | These grammar rules (besides `eof`) consume a single token of the kind
+    -- matching their name. The syntax node they produce has an "S.Token" kind,
+    -- unless otherwise specified.
+    lparen,
+    rparen,
+    lsquare,
+    rsquare,
+    lcurly,
+    rcurly,
+    comma,
+    colon,
+    kwImport,
+    kwFrom,
+    kwTrue,
+    kwFalse,
+    string,
+    int,
+    double,
+    lquote,
+    rquote,
+    quoted,
+    -- | A primitive for "T.Identifier" is not included: this should always be
+    -- done with an `as`, i.e. @"T.Identifier" ``as`` "S.DictKey"@
+    eof,
+
+    -- * Error Recovery
+
+    -- | This parser has built-in automatic error recovery. Currently, error
+    -- recovery occurs when an error is produced inside of a group rule. Recovery
+    -- is based on the possible tokens that can follow the current rule.
+    --
+    -- At a basic level, we can describe the mechanism as follows:
+    --
+    -- For a "GrammarRule" @label '<$$>' first '<>' next@, if an error occurs
+    -- while parsing @first@, check if the next token is in the "follow set" of
+    -- the group (the follow set of a rule is the set of tokens that can appear
+    -- directly after the rule in valid source). If it is, register an error in
+    -- the output and leave the token in the input. Otherwise, register an error and
+    -- discard the token.
+    --
+    -- As an example, consider the grammar rules:
+    --
+    -- @
+    -- stmts = eof <|> (stmt <> stmts)
+    -- stmt = Stmt <$$> "let" <> identifier <> "=" <> expr <> ";"
+    -- @
+    --
+    -- With the example source text:
+    --
+    -- @
+    -- let x = 2
+    -- let y = 4;
+    -- @
+    --
+    -- A parse error will occur at the beginning of line 2, when @let@ is found
+    -- instead of the expected @;@. The follow set of this @Stmt@ group is
```

Comment:
Ok if that is a name used in parsing theory!

Yes, let's maybe have a nice big explanation of what it is, if we don't yet.

Could we benefit from `follow token set` -> somehow adding `token` in there? Or is that redundant since you either know what follow set is or you don't? Yeah that could be so.